cli/azd/pkg/environment/environment.go

```diff
@@ -36,14 +37,78 @@ const ContainerRegistryEndpointEnvVarName = "AZURE_CONTAINER_REGISTRY_ENDPOINT"
 const ResourceGroupEnvVarName = "AZURE_RESOURCE_GROUP"
 
 type Environment struct {
-	// Values is a map of setting names to values.
-	Values map[string]string
+	// values is a map of setting names to values.
+	// all keys are converted to upper case before it is set.
+	values map[string]string
 	// File is a path to the file that backs this environment. If empty, the Environment
 	// will not be persisted when `Save` is called. This allows the zero value to be used
 	// for testing.
 	File string
 }
 
+// DeleteVariable removes an entry from azd environment which key matches the variable name.
+func (e *Environment) DeleteVariable(variableName string) {
+	delete(e.values, strings.ToUpper(variableName))
+}
+
+// SetVariable update or create an entry to the azd environment.
+// key is converted to upper case.
+func (e *Environment) SetVariable(key, value string) {
+	e.values[strings.ToUpper(key)] = value
+}
+
+// GetValue get the value of an entry with the requested key.
+// return empty string and false if the key is not found.
+// Use ValueOf if you don't need to check if value is found
+func (e *Environment) GetValue(key string) (string, bool) {
+	value, found := e.values[strings.ToUpper(key)]
+	return value, found
+}
+
+// GetValue get the value of an entry with the requested key.
+// return empty string and false if the key is not found.
+// Use GetValue to return if the value is found as well
+func (e *Environment) ValueOf(key string) string {
+	return e.values[strings.ToUpper(key)]
+}
+
+// HasValue return true if the key is found and the value is not empty string.
+func (e *Environment) HasValue(key string) bool {
+	value, found := e.values[strings.ToUpper(key)]
+	return found && value != ""
+}
+
+// Init make sure that values has an empty map.
+func (e *Environment) Init() {
+	if e.values == nil {
+		e.values = make(map[string]string)
+	}
+}
+
+// CopyValues get a copy of the azd environment variable entries.
+func (e *Environment) CopyValues() map[string]string {
+	if e.values == nil {
+		return nil
+	}
+
+	result := make(map[string]string)
```

Comment:
Could provide the length here to `make` for some initial memory allocation

Reply:
I though about it...  but, is a `map[string]string` a contagious memory struct?  
What would length mean for a `map[string]string` ?   

I guess providing the length would create the array with empty pointers to strings and allocate the string in the heap anyway for each key/value.

Anyway. I will add the length,  just wondering if it might be a super tiny difference for a `map[string]string`

Reply:
Yeah, I was curious myself initially. `map` is implemented as HashTable. It looks like this serves as a hint to calculate number of buckets? https://github.com/golang/go/blob/dfdf55158dcfc3ef1bd436b3b9ed6daa20801fdb/src/runtime/map.go#L304